### PR TITLE
Simplify namespace controller

### PR DESCRIPTION
**Pull request by lei-tang**
Please provide a description for what this PR is for: https://github.com/istio/istio/issues/22463

The current implementation of the namespace controller is complicated and difficult to reason its correctness. Simplifying the implementation makes code easier to reason its correctness and may solve the bugs in https://github.com/istio/istio/issues/22463. This PR makes the implementation very simple:
- The namespace controller only handles the events of adding a namespace and updating a namespace. The event handler simply writes the configmap to a namespace. The delete events are not handled because when the namespace is deleted, the configmap in the namespace is deleted too and there is no extra actions needed. The controller for configmap is removed to simplify the implementation. 
- The update handler of the namespace controller periodically writes the configmap to a namespace, which is useful when the namespace controller needs to sync the configmaps in namespaces. If only the adding events are handled, the namespace controller may miss the adding events and cause the configmap to be out-of-sync.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/bootstrap/namespacecontroller.go [hunk 1]
```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
 	getData func() map[string]string
 	client  corev1.CoreV1Interface
 
-	queue queue.Instance
-
 	// Controller and store for namespace objects
 	namespaceController cache.Controller
-	// Controller and store for ConfigMap objects
-	configMapController cache.Controller
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/namespacecontroller.go]: why is this removed? What happens if the config map is removed/changed?

File: pilot/pkg/bootstrap/namespacecontroller.go [hunk 2]
```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
...
+	namespaceStore      cache.Store
 }
 
 // NewNamespaceController returns a pointer to a newly constructed NamespaceController instance.
 func NewNamespaceController(data func() map[string]string, kubeClient kubernetes.Interface) *NamespaceController {
 	c := &NamespaceController{
 		getData: data,
 		client:  kubeClient.CoreV1(),
-		queue:   queue.NewQueue(time.Second),
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/namespacecontroller.go]: why are we removing the queue? This is standard across istio controllers to properly manage error handling, etc

File: pilot/pkg/bootstrap/namespacecontroller.go [hunk 3]
```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
...
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.client.Namespaces().Watch(options)
+			}}
 	})
-	c.namespaceController = namespaceInformer
+
+	c.namespaceStore, c.namespaceController =
+		cache.NewInformer(namespaceLW, &v1.Namespace{}, namespaceResyncPeriod, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.namespaceUpdated,
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/namespacecontroller.go]: why do we need to manage updates for a namespace?

File: pilot/pkg/bootstrap/namespacecontroller.go [hunk 4]
```diff
@@ -141,26 +135,3 @@ func (nc *NamespaceController) insertDataForNamespace(ns string) error {
 	}
 	return certutil.InsertDataToConfigMap(nc.client, meta, nc.getData())
 }
-
-// On namespace change, update the config map.
-// If terminating, this will be skipped
-func (nc *NamespaceController) namespaceChange(obj interface{}) error {
-	ns, ok := obj.(*v1.Namespace)
-
-	if ok && ns.Status.Phase != v1.NamespaceTerminating {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/namespacecontroller.go]: we are no longer handling ns termination. Will this lead to errors when a namespace is deleted?

File: pilot/pkg/bootstrap/namespacecontroller.go [hunk 5]
```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
...
 
+// When a namespace is created, Citadel adds its public CA certificate
+// to a well known configmap in the namespace.
+func (nc *NamespaceController) namespaceAdded(obj interface{}) {
+	ns, ok := obj.(*v1.Namespace)
+
+	if ok {
+		err := nc.insertDataForNamespace(ns.Name)
+		if err != nil {
+			log.Errorf("error when inserting CA cert to configmap: %v", err)
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/namespacecontroller.go]: error is not handled so any transient error will lead to permanently missing updates to namespace. That is why we had the queue

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 1 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go]: This is removed to simplify the implementation. If the config map is removed/changed, the namespace controller will periodically write the configmap again.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 2 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go]: The queue is removed to simplify the implementation. For example, the service controller in security/pkg/registry/kube/service.go and the service account controller in security/pkg/registry/kube/serviceaccount.go also do not use a queue.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 3 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go]: This update handler periodically writes the configmap to a namespace, which is useful when the controller needs to sync the configmaps in namespaces. If only adding events are handled, the controller may miss the adding events and cause the configmap to be out-of-sync. The following is an explanation of OnUpdate cited from go/pkg/mod/k8s.io/client-go@v0.18.0/tools/cache/controller.go.

//  OnUpdate is also called when a re-list happens, and it will
//      get called even if nothing changed. This is useful for periodically
//      evaluating or syncing something.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 5 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go]: Any transient error will be handled later by the update event handler.

**Comment by howardjohn** [comment]
#20915, which this PR is essentially reverting, was added to address performance issues with the old approach, which are now present. Namely that every 30s we are performing a write for every namespace, which can be 1000s of writes per minute. This impacts the rest of Pilot as well, as it shares a rate limiter, and has impact on the api server.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/bootstrap/namespacecontroller.go]: When a namespace is deleted the logs of Istiod will be spammed with a bunch of errors because we attempt to write to the deleted namespace

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 2 above)
Comment (ayj) [reply pilot/pkg/bootstrap/namespacecontroller.go]: The queue was added to rate limit change and allow work to be easily retried for transient failures. This is a common k8s controller pattern. See https://github.com/kubernetes/sample-controller/blob/master/controller.go

**Comment by ayj** [comment]
The controller could be simplified and improved, but I don't think reverting back is the way to do it. A few things stood out looking at the code:

1) The AddFunc handler for ConfigMaps is missing.
2) The AddFunc and DeleteFunc handlers for Namespace are missing.
3) Unit test coverage is lacking. One suggestion is to re-write the controller as a single reconcile loop / function that is called whenever a namespace/configmap change. Unit tests can populate the informer cache / fakeclient and invoke the reconcile function to verify state is reconciled. This doesn't test the informer logic itself or the Add/Update/Delete func handlers, but that is typically not where bugs exist (except in this case). The advantage of this approach is easier to write/read tests which are predictable and don't depend on Eventually/Consistently.  

See https://github.com/istio/istio/pull/22625 for some of the fixes for (1) and (2). I only ran the limited unit tests. I didn't test in-cluster.

**Comment by ayj** [comment]
I think a short term mitigation would be to add a resync period to the current code. Any missing configmap/namespace should be reconciled when everything is relisted.

**Comment by howardjohn** [comment]
I do have a PR to add resync by the way:
https://github.com/istio/istio/pull/22606

**Comment by ayj** [comment]
> I do have a PR to add resync by the way: #22606

You should be able to roll my PR into yours which would address https://github.com/istio/istio/pull/22606#issuecomment-606352742 I think.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 4 above)
Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go:151]: With this PR, the namespace controller only handles the events of adding a namespace and updating a namespace. Thus, when a namespace is deleted, there will be no error logs of attempting to write the configmap to the deleted namespace.

**Comment by lei-tang** [comment]
> The controller could be simplified and improved, but I don't think reverting back is the way to do it. A few things stood out looking at the code:
> 
> 1. The AddFunc handler for ConfigMaps is missing.
> 2. The AddFunc and DeleteFunc handlers for Namespace are missing.
> 3. Unit test coverage is lacking. One suggestion is to re-write the controller as a single reconcile loop / function that is called whenever a namespace/configmap change. Unit tests can populate the informer cache / fakeclient and invoke the reconcile function to verify state is reconciled. This doesn't test the informer logic itself or the Add/Update/Delete func handlers, but that is typically not where bugs exist (except in this case). The advantage of this approach is easier to write/read tests which are predictable and don't depend on Eventually/Consistently.
> 
> See #22625 for some of the fixes for (1) and (2). I only ran the limited unit tests. I didn't test in-cluster.

This PR simplifies the namespace controller implementation:
- The namespace controller only handles the events of adding a namespace and updating a namespace. The event handler simply writes the configmap to a namespace. The controller for configmap is removed to simplify the implementation. 
- The update handler of the namespace controller periodically writes the configmap to a namespace, which is useful when the namespace controller needs to sync the configmaps in namespaces. If only the adding events are handled, the namespace controller may miss the adding events and cause the configmap to be out-of-sync.

File: pilot/pkg/bootstrap/namespacecontroller.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/bootstrap/namespacecontroller.go:151]: Lei, I have deployed this pr to a cluster and tried it. A delete triggers the update callback as well

Comment (lei-tang) [reply pilot/pkg/bootstrap/namespacecontroller.go:151]: Thanks! A check of namespace terminating is added as follows:
if ok && ns.Status.Phase != v1.NamespaceTerminating {
...
}

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 6]
```diff
@@ -52,88 +50,77 @@ type NamespaceController struct {
...
-			c.queue.Push(func() error {
-				return c.namespaceChange(obj)
-			})
-		},
-		UpdateFunc: func(_, obj interface{}) {
-			c.queue.Push(func() error {
-				return c.namespaceChange(obj)
-			})
-		},
+	namespaceLW := listwatch.MultiNamespaceListerWatcher([]string{metav1.NamespaceAll}, func(namespace string) cache.ListerWatcher {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:65]: Why change to this?

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 7]
```diff
@@ -52,88 +50,77 @@ type NamespaceController struct {
...
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.client.Namespaces().Watch(context.TODO(), options)
+			}}
 	})
-	c.namespaceController = namespaceInformer
+
+	c.namespaceStore, c.namespaceController =
+		cache.NewInformer(namespaceLW, &v1.Namespace{}, NamespaceResyncPeriod, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.namespaceUpdated,
+			AddFunc:    c.namespaceAdded,
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:78]: you do this synchronously in the callback, if this takes long, it will block list-watch.

It is better to put back the queue.

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 8]
```diff
@@ -52,88 +50,77 @@ type NamespaceController struct {
...
 
 // NewNamespaceController returns a pointer to a newly constructed NamespaceController instance.
 func NewNamespaceController(data func() map[string]string, options Options, kubeClient kubernetes.Interface) *NamespaceController {
 	c := &NamespaceController{
 		getData: data,
 		client:  kubeClient.CoreV1(),
-		queue:   queue.NewQueue(time.Second),
 	}
 
-	configmapInformer := informer.NewFilteredConfigMapInformer(kubeClient, metav1.NamespaceAll, options.ResyncPeriod,
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:71]: What will happen if the cm id deleted? It loses the way to be aware of it.

**Comment by lei-tang** [comment]
Based on the review comments, this PR is closed. Anyone interested in trying this PR may download it and create a local Istio build including it.
